Fix build with upstream gtk

marmistrz opened the pull request:
Fixes #6.

Not all features are available with upstream gtk2, but the apps should mostly work (unless they are directly using the Maemo Gtk+ API). To be more precise, the following libhildon features are missing:
* placeholders for text fields
* Hildon IM mode features
* Hildon theming features (setting the UI size)

marmistrz committed:
Fix inline functions not linking properly.

In C, bare inline functions are alternative to external ones. This
causes linking problems on some systems. Those functions are clearly
internal and have been marked as `static inline`.

See this stackoverflow post for more details:
https://stackoverflow.com/questions/16245521/c99-inline-function-in-c-file/16245669

MerlijnWajer commented on hildon/hildon-live-search.c:
```diff
@@ -761,11 +760,13 @@ hildon_live_search_init                         (HildonLiveSearch *self)
     /* Unset the autocap and dictionary input flags from the
        HildonEntry. */
-    imode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
+    HildonGtkInputMode    imode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
```
Can we keep the def of the var at the top, in an ifdef? I think that's better C practice, is it not?

marmistrz replied:
I'd disagree it's a better practice. It comes from the times of ANSI C (C90), where you had to place all the variable definitions at the top. Since C99 it's permitted to define them everywhere.

I'd prefer not to have unused variables in my code. It's error prone (it's an uninitialized variable which I can unwittingly pass somewhere else and trigger UB) and will trigger warnings with correct compiler options. I know that libhildon currently has over 700 warnings, but I'd prefer not to add more of them. There's no performance advantage, since the resulting assembly will most likely be the same anyway. This way I'm sure that I won't use this variable unless I use the Maemo Gtk+ API.

If you insist, I can place it at the top at the cost of adding another ifdef, but I don't think it's a clean solution.

marmistrz replied:
Done.

MerlijnWajer commented on hildon/hildon-touch-selector-entry.c:
```diff
@@ -239,22 +239,23 @@ hildon_touch_selector_entry_init (HildonTouchSelectorEntry *self)
   gtk_entry_set_activates_default (GTK_ENTRY (priv->entry), TRUE);
-  input_mode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
+#ifdef MAEMO_GTK
+  HildonGtkInputMode input_mode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
```
Same as other comment

MerlijnWajer commented:
On 06/01/2020 21:37, Marcin Mielniczuk wrote:
> marmistrz commented on this pull request.
> 
> 
> 
>>      /* Unset the autocap and dictionary input flags from the
>         HildonEntry. */
> -    imode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
> +    HildonGtkInputMode    imode = hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
> 
> I'd disagree it's a better practice. It comes from the times of ANSI C (C90), where you had to place all the variable definitions at the top. Since C99 it's permitted to define them everywhere.
> 
> I'd prefer not to have unused variable in my code. It's error prone (it's an uninitialized variable which I can unwittingly pass somewhere else and trigger UB) and will trigger warnings with correct compiler options. I know that libhildon currently has over 700 warnings, but I'd prefer not to add more of them. There's no performance advantage, since the resulting assembly will most likely be the same anyway. This way I'm sure that I won't use this variable unless I use the Maemo Gtk+ API.

I think with a second ifdef, the variable would not be unused, right? It
would just not be defined.

Are you sure all the warnings you are seeing are also present when
building against patched gtk?

marmistrz commented:
> Are you sure all the warnings you are seeing are also present when building against patched gtk?

I don't know. I only see a ton of warnings on my host Arch Linux. I found the following types of them
* deprecation warnings: mostly `GTimeVal` -> `GTimeDate`, `GTypeDebugFlags`.
* `G_ADD_PRIVATE` - I don't even know how to silence it.
* unused but set variables
* `-Wmaybe-uninitialized`

MerlijnWajer commented:
This is probably for another issue, but you might be building against another gtk version.

marmistrz commented:
I'm building against 2.24.32-1, Leste seems to have 2.24.25.

spinal84 commented on hildon/hildon-gtk.h:
```diff
@@ -66,6 +66,18 @@ typedef enum {
+
+// Support for placeholder texts is used extensively around libhildon,
+// but requires a patched version of Gtk+. It was added upstream in Gtk 3.2.
+// Therefore, for a non-Maemo version of Gtk+ we expose a dummy function.
```
Please try to keep the style of the package you're making changes to.
This package have comments of the form /* */ in most cases.

marmistrz replied:
Done.

spinal84 commented on hildon/hildon-gtk.c:
```diff
@@ -31,6 +31,24 @@
 #include "hildon-edit-toolbar-private.h"
 #include "hildon-private.h"
 
+// Add placeholder implementations for Maemo-specific Gtk functions
```
Please try to keep the style of the package you're making changes to.
This package have comments of the form /* */ in most cases.

spinal84 replied:
This comment seems to be a good commit message.
I don't think it's really needed here, in the code.

marmistrz replied:
It's useful in the code - you don't usually look at the commit history while analyzing the code.

marmistrz replied:
And I changed the comment style

spinal84 commented on examples/Makefile.am:
```diff
@@ -45,8 +43,10 @@ EXAMPLES				= hildon-window-example				\
+MAEMO_GTK_EXAMPLES  = hildon-pannable-area-touch-list-example	\
+					  hildon-pannable-area-touch-grid-example \
+					  hildon-live-search-example			\
+					  hildon-live-search-grid-example
```
Changing tab/spaces count here makes the code look inconsistent. Please fix.

marmistrz replied:
I didn't change to spaces. It's neatly aligned with 4 spaces tab. But I made it well-aligned with 8-space tabs. (This is why one should intend with spaces! :))

spinal84 commented on hildon/hildon-entry.c:
```diff
@@ -71,7 +71,11 @@ set_property                                    (GObject      *object,
 	  size |= HILDON_SIZE_FINGER_HEIGHT;
-	hildon_gtk_widget_set_theme_size (GTK_WIDGET (object), size);
+#ifdef MAEMO_GTK
+        hildon_gtk_widget_set_theme_size (GTK_WIDGET (object), size);
```
Changing initial tab to spaces here seems redundant.
The commits history should reflect code changes, not your editor preferences.

marmistrz replied:
This was not tab -> spaces. For some reason I thought that this code was misindented, but it's not the case. I'll revert this one.

spinal84 replied:
I can only see tab changed to 8 spaces here.

marmistrz replied:
The code around here had a messy mixture of tabs&spaces. I changed everything to spaces for consistency.

spinal84 commented on hildon/hildon-gtk.h:
```diff
@@ -18,6 +18,55 @@
 
 #include                                        <gtk/gtk.h>
+/// Temporary hack to provide the types from the Maemo fork of gtk
+/// This was copy-pasted from hildonize-gtk-enums.diff from maemo-leste/gtk
```
Pull requests should be complete.
Please try to avoid hacks, temporary hacks, dirty hacks and so on in your pull requests.
If this part of file is going to be changed, please implement the needed changes in your pull request. If it's not going to be changed, then remove this comment.

marmistrz replied:
This used to be a temporary hack while this was only a partial fix. But you reminded me that I had intended to move it to a separate file, which I've just done.

spinal84 commented on hildon/hildon-gtk.h:
```diff
@@ -18,6 +18,55 @@
+  HILDON_GTK_INPUT_MODE_DICTIONARY        = 1 << 31
+} HildonGtkInputMode;
+
+/* Temporary compatibility define */
```
Same.

marmistrz replied:
This is not my comment, it was taken directly from [this patch](https://github.com/maemo-leste/gtk/blob/552c94ce4a423dfc5fe775dd71f55388be7b2d8d/debian/patches/hildonize-gtk-enums.diff#L28). If you remove this comment from maemo-leste/gtk, I'll remove it from here.

spinal84 replied:
OK, let's leave it then.

marmistrz committed:
Fix build with upstream Gtk+.

This commit disables the portions of code using the Maemo Gtk+ API when building
against stock Gtk+. Not all features will be available, but programs not
using the Maemo Gtk+ API should mostly run.

Furthermore, placeholder no-op implementations of the following Maemo
Gtk+ API function will be provided in such case, precisely:
* hildon_gtk_entry_set_placeholder_text
* hildon_gtk_text_view_set_placeholder_text

MerlijnWajer commented:
@spinal84 - it looks good to me. If  you think it's OK too, let's merge it.

spinal84 commented on hildon/hildon-color-chooser.c:
```diff
@@ -109,19 +109,19 @@ init_borders                                    (GtkWidget *w,
                                                  GtkBorder *outer);
 
-inline void 
+static inline void 
```
Please remove the trailing spaces in the lines you modify in this file.
It's not your fault, but if you touch these lines, it's better to be fixed.

_Hint: inspecting your commits with "git show" is always a good idea._

marmistrz replied:
I'll just sanitize all trailing whitespaces in the whole file, ok?

spinal84 replied:
If you do it in separate commit, it would be fine.

marmistrz replied:
Will do, once everything else is fine.

spinal84 replied:
Ok. Please let me know, when you are ready for a new review.

spinal84 commented on hildon/hildon-entry.c:
```diff
@@ -67,11 +67,15 @@ set_property                                    (GObject      *object,
+        size = g_value_get_flags (value);
+        /* If this is auto height, default to finger height. */
+        if (!(size & (HILDON_SIZE_FINGER_HEIGHT | HILDON_SIZE_THUMB_HEIGHT)))
+            size |= HILDON_SIZE_FINGER_HEIGHT;
```
Please, move such style-fixing changes to a separate commit to distinguish them from code changes.

In this example the commit message may be: "hildon/hildon-entry.c: fix tabs/spaces mess"

If several files are changed, just omit the filename part from the commit message.

marmistrz replied:
This are minor and time-consuming changes. Do we really have to complicate our lives that much for such minor changes?

spinal84 replied:
Nice commit history is really appreciated when someone wants to hack on the code.
"git blame" shouldn't confuse users.

I understand, why you don't want to do the splitting/rebasing commits, I did it myself many times. It's not that hard as it seems to be.

spinal84 commented on hildon/hildon-entry.c:
```diff
@@ -67,11 +67,15 @@ set_property                                    (GObject      *object,
+        if (!(size & (HILDON_SIZE_FINGER_HEIGHT | HILDON_SIZE_THUMB_HEIGHT)))
+            size |= HILDON_SIZE_FINGER_HEIGHT;
+#ifdef MAEMO_GTK
+        hildon_gtk_widget_set_theme_size (GTK_WIDGET (object), size);
```
This line should remain unchanged in this commit.
Please, add it to style-fixing commit mentioned above.

spinal84 commented on hildon/hildon-gtk.c:
```diff
@@ -31,6 +31,24 @@
+{
+    g_warning("hildon_gtk_text_view_set_placeholder_text unsupported");
+}
+
```
Why is this empty line needed?

marmistrz replied:
Fixed.

spinal84 commented on hildon/hildon-gtk.c:
```diff
@@ -31,6 +31,24 @@
+#ifndef MAEMO_GTK
+void
+hildon_gtk_entry_set_placeholder_text           (GtkEntry *entry,
+                                                 const gchar *text)
```
Arguments are not aligned, please fix.
Follow the style of other functions from this file.

marmistrz replied:
Fixed

spinal84 commented on hildon/hildon-gtk.c:
```diff
@@ -31,6 +31,24 @@
+
+void
+hildon_gtk_text_view_set_placeholder_text       (GtkEntry *text_view,
+                                                 const gchar    *text)
```
Arguments are not aligned, please fix.
Follow the style of other functions from this file.

marmistrz replied:
Moved to a different file and fixed

spinal84 commented on hildon/hildon-gtk.c:
```diff
@@ -31,6 +31,24 @@
 #include "hildon-edit-toolbar-private.h"
 #include "hildon-private.h"
 
+/* Add placeholder implementations for Maemo-specific Gtk functions */
```
This comment seems excessive. It's to obvious what is happening here just looking at the code.

Adding comments for obvious things is not a good idea. But if you really think it's needed here, ok.

Just a recommendation.

marmistrz replied:
Moved to a different file

spinal84 commented on hildon/hildon-touch-selector-entry.c:
```diff
@@ -424,6 +429,9 @@ hildon_touch_selector_entry_set_input_mode (HildonTouchSelectorEntry * selector,
 
   hildon_gtk_entry_set_input_mode (GTK_ENTRY (priv->entry), input_mode);
+#else
+  g_warning("hildon_gtk_entry_set_input_mode unsupported");
```
This warning differs from others. Please keep the same style. I mean "no-op" part.
Either remove it or use it the same way everywhere.

marmistrz replied:
Fixed

spinal84 commented on hildon/hildon-touch-selector-entry.c:
```diff
@@ -440,13 +448,18 @@ hildon_touch_selector_entry_set_input_mode (HildonTouchSelectorEntry * selector,
 
   return hildon_gtk_entry_get_input_mode (GTK_ENTRY (priv->entry));
+#else
+  g_warning("hildon_touch_selector_entry_get_input_mode: no-op, hildon_gtk_entry_get_input_mode unsupported");
```
This warning differs from others. Please keep the same style. I mean the comma after "no-op" part.

marmistrz replied:
Fixed

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,66 @@
+#define MAEMOGTKCOMPAT_H
+
+// Provides the types from the Maemo Gtk+ API if upstream Gtk+ is used.
+// This was copy-pasted from hildonize-gtk-enums.diff from maemo-leste/gtk
```
Please use multiline style comments (/* */).

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,66 @@
+
+/* Support for placeholder texts is used extensively around libhildon,
+but requires a patched version of Gtk+. It was added upstream in Gtk 3.2.
+Therefore, for a non-Maemo version of Gtk+ we expose dummy functions. */
```
Please, align the 2nd and the 3rd lines in this comment.
Check other multiline comments from the package for a guideline.

marmistrz replied:
fixed

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,66 @@
+#ifndef MAEMOGTKCOMPAT_H
```
Please add the copyright header to this file.
Check other headers from the source for a guide.

marmistrz replied:
fixed

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,66 @@
+#ifndef MAEMOGTKCOMPAT_H
+#define MAEMOGTKCOMPAT_H
```
Try to keep the style of the package you are making changes to.
`#define __MAEMOGTKCOMPAT_H__`
... will be a more consistent choice.

marmistrz replied:
fxied

spinal84 commented:
1. Your maemo-gtk-compat.c is not going to be built (it's missing in Makefile).
2. Please add an entry to debian/changelog. Use "2.2.26-1" as a version number.

marmistrz committed:
hildon/hildon-color-chooser.c: Sanitize trailing whitespace

marmistrz committed:
hildon/hildon-entry.c: Sanitize tabs & spaces

marmistrz committed:
Update debian/changelog (2.2.26-1)

marmistrz commented:
This shou'd be ready for review again now.

> Your maemo-gtk-compat.c is not going to be built (it's missing in Makefile).

Ouch! It must have gotten lost between force resets.

> Please add an entry to debian/changelog. Use "2.2.26-1" as a version number.

Done.

Btw. @spinal84 how can I enable tests in `configure`? There seems to be no option for that?

spinal84 commented:
> how can I enable tests in `configure`? There seems to be no option for that?
```
# apt install check; ./configure | grep test
config.status: creating tests/Makefile
- Build unit tests...: yes
```

spinal84 commented:
Please use "Update debian/changelog (2.2.26-1)" as the last commit message (for the debian changelog update).

marmistrz commented:
Done.

spinal84 commented:
Please fix all the warnings and errors introduced by your commits.
https://pastebin.com/5PVrEYE6

_Tip: don't do pull requests for untested changes._

marmistrz commented:
@spinal84 I tested them locally on my desktop Arch Linux and the change compiles successfully there, against gtk 2.24.32-1. 
That's the purpose of CI to check if I'm not inadvertently breaking any other platform.

I'll see if I can reproduce the errors in the Leste chroot.

marmistrz commented:
@spinal84 I can reproduce and see what the issue is. I thought that `MAEMO_GTK` is simply defined in some header file but it's actually just a flag passed to the C compiler. Therefore, if any other application includes `hildon-gtk.h`, it will use the compatibility headers, which it should not.

Would it be possible to generate some header file using autotools so that the `MAEMO_GTK` define will be persistently present, even after installation, provided that `libhildon` was configured with Maemo Gtk+ API?

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,88 @@
+
+#endif
+
+#endif // __MAEMOGTKCOMPAT_H__
```
Please, use multiline style comment here.
Check other headers for examples.

spinal84 commented on hildon/maemo-gtk-compat.h:
```diff
@@ -0,0 +1,88 @@
+
+void
+hildon_gtk_text_view_set_placeholder_text       (GtkEntry *text_view,
+                                                 const gchar    *text);
```
Args not aligned

spinal84 commented:
> @spinal84 I can reproduce and see what the issue is. I thought that `MAEMO_GTK` is simply defined in some header file but it's actually just a flag passed to the C compiler. Therefore, if any other application includes `hildon-gtk.h`, it will use the compatibility headers, which it should not.
> 
> Would it be possible to generate some header file using autotools so that the `MAEMO_GTK` define will be persistently present, even after installation, provided that `libhildon` was configured with Maemo Gtk+ API?

It's possible. But it seems to be a wrong way. The proper way would be to recover the define which Fremantle used for gtk+-2. Look at this /usr/lib/pkgconfig/gtk+-2.0.pc:
`Cflags: -I${includedir}/gtk-2.0  -DMAEMO_CHANGES`

Compare to Leste /usr/lib/x86_64-linux-gnu/pkgconfig/gtk+-2.0.pc:
`Cflags: -I${includedir}/gtk-2.0`

The define was dropped by our developers who ported gtk2+ from Fremantle to Leste. They probably decided not to support stock gtk which Fremantle seemed to do. Because it needs a lot of work and time to be implemented properly.

I will make a pull request to restore it.

spinal84 commented:
Forget what I just wrote. I overcomplicated things.

There's no need to change maemo-leste/gtk package to fix the issues you have. MAEMO_GTK is defined using pkgconfig file installed by libhildon.

I'm going to make a PR to your repository with the needed changes. Please integrate them to your work and update this PR accordingly.

spinal84 committed:
hildon/hildon-gtk.h: remove executable bit

spinal84 committed:
Minor style fixes.

spinal84 committed:
Fix package building with the latest changes

spinal84 commented:
https://github.com/marmistrz/libhildon/pull/1

marmistrz commented:
Thanks a lot! I confirm that this works for me locally and also `dpkg-buildpackage` works in the leste vm.

MerlijnWajer commented:
Once this is ready, let's get it merged -- it's blocking `libhildon` from being built for `beowulf` / `buster`, and in turn blocks `libhildondesktop` from being built, blocking `maemo-launcher`, etc. :) 

https://phoenix.maemo.org/job/libhildon-binaries/lastBuild/architecture=amd64,label=amd64/consoleText

MerlijnWajer commented:
Also - thanks for the work both of you. Sometimes this can be tedious to get right.

marmistrz commented:
I'll rebase in the evening CET.

marmistrz commented:
Done.

marmistrz commented:
Btw. you guys should really consider setting up a formatter for libhildon & family, it's a waste of time to fix all this formatting manually, especially with these non-standard conventions.

spinal84 commented:
> Btw. you guys should really consider setting up a formatter for libhildon & family, it's a waste of time to fix all this formatting manually, especially with these non-standard conventions.

It's a common practice to follow existing formatting in community developed packages for integrity.

From https://github.com/maemo-leste/libhildon/blob/master/HACKING:
`* Obey the rules present in the existing code base -- it's pretty consistent`

P.S. Regarding the “formatter setup for libhildon & family”, you can share your “formatter setup” approach on the Maemo Leste wiki or IRC. I think that many people will be grateful for the exchange of effective programming skills.

marmistrz commented:
Of course it's a common practice. I'm just saying that it's much better and time-efficient if this practice is automated.

Of course, I could share my clang-format configs, but they don't adhere to the conventions in here. The tricky part is to setup the config so that it doesn't trigger a mass reformat.

MerlijnWajer commented:
I think this is causing trouble for me when building libhildondesktop:

```
/bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../..  -DMAEMO_CHANGES -DMAEMO_GTK -pthread -I/usr/include/hildon-1 -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -Wno-error=unused-but-set-variable -Wall -Wmissing-prototypes -Wmissing-declarations -Werror -Wno-format-extra-args -DG_DISABLE_CAST_CHECKS -c -o example_label_home_applet_la-example-label-home-applet.lo `test -f 'example-label-home-applet.c' || echo './'`example-label-home-applet.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -DMAEMO_CHANGES -DMAEMO_GTK -pthread -I/usr/include/hildon-1 -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -Wno-error=unused-but-set-variable -Wall -Wmissing-prototypes -Wmissing-declarations -Werror -Wno-format-extra-args -DG_DISABLE_CAST_CHECKS -c example-label-home-applet.c  -fPIC -DPIC -o .libs/example_label_home_applet_la-example-label-home-applet.o
In file included from /usr/include/hildon-1/hildon/hildon-caption.h:30,
                 from /usr/include/hildon-1/hildon/hildon.h:31,
                 from example-label-home-applet.c:27:
/usr/include/hildon-1/hildon/hildon-enum-types.h:60: error: "HILDON_TYPE_GTK_INPUT_MODE" redefined [-Werror]
 #define HILDON_TYPE_GTK_INPUT_MODE (hildon_gtk_input_mode_get_type())

In file included from /usr/include/gtk-2.0/gtk/gtktypeutils.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkobject.h:37,
                 from /usr/include/gtk-2.0/gtk/gtkwidget.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkcontainer.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkbin.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkwindow.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkdialog.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkaboutdialog.h:32,
                 from /usr/include/gtk-2.0/gtk/gtk.h:33,
                 from example-label-home-applet.c:26:
/usr/include/gtk-2.0/gtk/gtktypebuiltins.h:195: note: this is the location of the previous definition
 #define HILDON_TYPE_GTK_INPUT_MODE (hildon_gtk_input_mode_get_type ())

In file included from /usr/include/hildon-1/hildon/hildon-caption.h:30,
                 from /usr/include/hildon-1/hildon/hildon.h:31,
                 from example-label-home-applet.c:27:
/usr/include/hildon-1/hildon/hildon-enum-types.h:62: error: "HILDON_TYPE_MODE" redefined
-Werror]
 #define HILDON_TYPE_MODE (hildon_mode_get_type())

In file included from /usr/include/gtk-2.0/gtk/gtktypeutils.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkobject.h:37,
                 from /usr/include/gtk-2.0/gtk/gtkwidget.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkcontainer.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkbin.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkwindow.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkdialog.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkaboutdialog.h:32,
                 from /usr/include/gtk-2.0/gtk/gtk.h:33,
                 from example-label-home-applet.c:26:
/usr/include/gtk-2.0/gtk/gtktypebuiltins.h:197: note: this is the location of the previous definition
 #define HILDON_TYPE_MODE (hildon_mode_get_type ())

In file included from /usr/include/hildon-1/hildon/hildon-caption.h:30,
                 from /usr/include/hildon-1/hildon/hildon.h:31,
                 from example-label-home-applet.c:27:
/usr/include/hildon-1/hildon/hildon-enum-types.h:64: error: "HILDON_TYPE_UI_MODE" redefined [-Werror]
 #define HILDON_TYPE_UI_MODE (hildon_ui_mode_get_type())

In file included from /usr/include/gtk-2.0/gtk/gtktypeutils.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkobject.h:37,
                 from /usr/include/gtk-2.0/gtk/gtkwidget.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkcontainer.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkbin.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkwindow.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkdialog.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkaboutdialog.h:32,
                 from /usr/include/gtk-2.0/gtk/gtk.h:33,
                 from example-label-home-applet.c:26:
/usr/include/gtk-2.0/gtk/gtktypebuiltins.h:199: note: this is the location of the previous definition
 #define HILDON_TYPE_UI_MODE (hildon_ui_mode_get_type ())

In file included from /usr/include/hildon-1/hildon/hildon-caption.h:30,
                 from /usr/include/hildon-1/hildon/hildon.h:31,
                 from example-label-home-applet.c:27:
/usr/include/hildon-1/hildon/hildon-enum-types.h:66: error: "HILDON_TYPE_SIZE_TYPE" redefined [-Werror]
 #define HILDON_TYPE_SIZE_TYPE (hildon_size_type_get_type())

In file included from /usr/include/gtk-2.0/gtk/gtktypeutils.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkobject.h:37,
                 from /usr/include/gtk-2.0/gtk/gtkwidget.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkcontainer.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkbin.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkwindow.h:36,
                 from /usr/include/gtk-2.0/gtk/gtkdialog.h:35,
                 from /usr/include/gtk-2.0/gtk/gtkaboutdialog.h:32,
                 from /usr/include/gtk-2.0/gtk/gtk.h:33,
                 from example-label-home-applet.c:26:
/usr/include/gtk-2.0/gtk/gtktypebuiltins.h:201: note: this is the location of the previous definition
 #define HILDON_TYPE_SIZE_TYPE (hildon_size_type_get_type ())

cc1: all warnings being treated as errors
make[4]: *** [Makefile:512: example_label_home_applet_la-example-label-home-applet.lo] Error 1
make[4]: Leaving directory '/root/libhildondesktop/examples/home'
make[3]: *** [Makefile:397: all-recursive] Error 1
make[3]: Leaving directory '/root/libhildondesktop/examples'
make[2]: *** [Makefile:444: all-recursive] Error 1
make[2]: Leaving directory '/root/libhildondesktop'
make[1]: *** [Makefile:376: all] Error 2
make[1]: Leaving directory '/root/libhildondesktop'
dh_auto_build: make -j2 returned exit code 2
make: *** [debian/rules:3: build] Error 2
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```

MerlijnWajer commented:
Reverting `a5b9dfe71c31058b7f7144bfd557bb178596b586` fixes the problem for me.

MerlijnWajer commented:
I've reverted above commit for my bayamo/beowulf tag because it breaks all libhildon users.

marmistrz commented:
~@MerlijnWajer I see what the problem is: this commit by @spinal84 ac53ca26bf8b99cee795cfb9c6a3983ddd0b7c4d doesn't actually do the right thing: It only defines  `MAEMO_GTK` for the purpose of gtk-doc but the `MAEMO_GTK` define is still missing whenever a different package is using libhildon later on. I must have missed that.~

~We probably should either define `MAEMO_GTK` in the leste fork of `gtk2` or put it into `hildon-gtk.h` whenever libhildon is configured to use the Maemo Gtk+. (some automake-based-autogeneration)~

~Probably the former is a better idea.~

spinal84 commented:
@marmistrz, be careful, please. Do you see that magic "-DMAEMO_CHANGES -DMAEMO_GTK" part here?
```
/bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../..  -DMAEMO_CHANGES -DMAEMO_GTK -pthread -I/usr/include/hildon-1 -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   -Wno-error=unused-but-set-variable -Wall -Wmissing-prototypes -Wmissing-declarations -Werror -Wno-format-extra-args -DG_DISABLE_CAST_CHECKS -c -o example_label_home_applet_la-example-label-home-applet.lo `test -f 'example-label-home-applet.c' || echo './'`example-label-home-applet.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../.. -DMAEMO_CHANGES -DMAEMO_GTK -pthread -I/usr/include/hildon-1 -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -Wno-error=unused-but-set-variable -Wall -Wmissing-prototypes -Wmissing-declarations -Werror -Wno-format-extra-args -DG_DISABLE_CAST_CHECKS -c example-label-home-applet.c  -fPIC -DPIC -o .libs/example_label_home_applet_la-example-label-home-applet.o
In file included from /usr/include/hildon-1/hildon/hildon-caption.h:30,
                 from /usr/include/hildon-1/hildon/hildon.h:31,
                 from example-label-home-applet.c:27:
/usr/include/hildon-1/hildon/hildon-enum-types.h:60: error: "HILDON_TYPE_GTK_INPUT_MODE" redefined [-Werror]
 #define HILDON_TYPE_GTK_INPUT_MODE (hildon_gtk_input_mode_get_type())
```

marmistrz commented:
@spinal84, you're right, sorry!

spinal84 commented:
The problem is treating warnings as errors:
`cc1: all warnings being treated as errors`

Treating warnings as errors is a very good practice, and I don't advise changing this option here just for workaround. The warnings should be fixed.

Please, read above error messages carefully and fix the root of the issue.
I'm reopening the pull request.

MerlijnWajer commented:
Are you sure that's the problem? I think the problem seems to be redefinitions in the file that was added - which should probably not be there at all for us.

spinal84 commented:
@marmistrz, I rebased your commits before merging: there were more issues and didn't want to waste more time on pointing and fixing every issue by you.

You will have to fetch maemo-leste/libhildon/master to resolve your branch conflicts.
You may open a new PR if you have problems updating this one.

spinal84 commented:
@MerlijnWajer, yes, I saw these warnings earlier.
When building the deb package with the pull request commits applied.

marmistrz commented:
The redefinition of `HILDON_TYPE_MODE` occurs in `/usr/include/gtk-2.0/gtk/gtktypebuiltins.h:197` and `hildon/hildon-enum-types.h:62` and neither of files were touched by my changes. This looks like an issue with your codebase. 

It is apparently first defined here: (that's the only place where it appears in gtk).
https://github.com/maemo-leste/gtk/blob/bf9a7dfb8f698ef900bf21182d97785ac50f1ddf/debian/patches/hildonize-gtk-widget.diff#L270

and later put into `hildon-enum-types.h`. I have no idea why you're doing that, but it seems completely wrong and it's not my fault that you're redefining stuff and at some point it blows up.

My change did not add `HILDON_TYPE_MODE`.

MerlijnWajer commented:
The additions from `maemo-gtk-compat.h` to `hildon-enum-types.h` are where the re-definitions come from. One of the files wasn't there before, I believe.

Why don't you build both in a VM and see?

In any case, I'm reverting this locally so I can move forward with the important bits, and not deal with breakage over a usecase we just don't intend to support. Please try to get this fixed up.

MerlijnWajer commented:
```
# diff /usr/include/hildon-1/hildon/hildon-enum-types.h /tmp/hildon-enum-types.h
55a56,66
> /* Enumerations from "maemo-gtk-compat.h" */
> GType gtk_invalid_input_type_get_type (void);
> #define HILDON_TYPE_INVALID_INPUT_TYPE (gtk_invalid_input_type_get_type())
> GType hildon_gtk_input_mode_get_type (void);
> #define HILDON_TYPE_GTK_INPUT_MODE (hildon_gtk_input_mode_get_type())
> GType hildon_mode_get_type (void);
> #define HILDON_TYPE_MODE (hildon_mode_get_type())
> GType hildon_ui_mode_get_type (void);
> #define HILDON_TYPE_UI_MODE (hildon_ui_mode_get_type())
> GType hildon_size_type_get_type (void);
> #define HILDON_TYPE_SIZE_TYPE (hildon_size_type_get_type())
```

spinal84 commented:
@marmistrz, again, please be careful.
hildon/hildon-enum-types.h is generated automatically from the sources.
You have to dig deeper to fix the issue.

marmistrz commented:
> ```
> # diff /usr/include/hildon-1/hildon/hildon-enum-types.h /tmp/hildon-enum-types.h
> 55a56,66
> > /* Enumerations from "maemo-gtk-compat.h" */
> > GType gtk_invalid_input_type_get_type (void);
> > #define HILDON_TYPE_INVALID_INPUT_TYPE (gtk_invalid_input_type_get_type())
> > GType hildon_gtk_input_mode_get_type (void);
> > #define HILDON_TYPE_GTK_INPUT_MODE (hildon_gtk_input_mode_get_type())
> > GType hildon_mode_get_type (void);
> > #define HILDON_TYPE_MODE (hildon_mode_get_type())
> > GType hildon_ui_mode_get_type (void);
> > #define HILDON_TYPE_UI_MODE (hildon_ui_mode_get_type())
> > GType hildon_size_type_get_type (void);
> > #define HILDON_TYPE_SIZE_TYPE (hildon_size_type_get_type())
> ```

That's weird, because `HILDON_TYPE_MODE` is present exclusively in gtk. I'll take a look when I have some time, but this probably won't be until late February.

spinal84 commented:
Closed by https://github.com/maemo-leste/libhildon/pull/8
